retworkx-core/src/shortest_path/bellman_ford.rs

```diff
@@ -0,0 +1,392 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use std::collections::VecDeque;
+use std::hash::Hash;
+
+use fixedbitset::FixedBitSet;
+use petgraph::algo::kosaraju_scc;
+use petgraph::algo::{is_cyclic_directed, Measure};
+use petgraph::graph::IndexType;
+use petgraph::stable_graph::StableDiGraph;
+use petgraph::visit::{
+    EdgeRef, IntoEdges, IntoNodeIdentifiers, NodeCount, NodeIndexable, Visitable,
+};
+
+use crate::dictmap::*;
+use crate::distancemap::DistanceMap;
+
+struct BellmanFordData<G, S>
+where
+    G: IntoEdges + Visitable + NodeIndexable + NodeCount + IntoNodeIdentifiers,
```

Comment:
I always like to use only the required traits.
```suggestion
    G: petgraph::visit::GraphBase,
```